### ISSUE TITLE
`ruff rule` should include a property that indicates the fix is unsafe or not

### ISSUE THREAD

**Issue by zender-vivodyne** [astral-sh/ruff#13108]
## Usecase

We have a legacy python project that we want to ease into using ruff and would like to start with just formatting and any linting that  can be automatically fixed. Using the `ruff rule --all --output-format json` we should be able to build a command that allows for that. 

Command we are using to build the list (tested only on MacOS)

```
ruff rule --all --output-format json | jq -r '.[] | select(.fix == "Fix is always available." and .preview == false and .code ) | .code' | sort | paste -sd, - | sed 's/[^,]*/"&"/g'
```

We can then just take this and put it into the `select` section of the config. 

Since there is no flag it requires us to copy and paste the above, then run `ruff check --fix --statistics` and copy all the values that are reported after not being fixed into the `ignore` list. 

## Current state
`ruff rule --all --output-format json`

Example output:

```
[
...
{
    "name": "verbose-raise",
    "code": "TRY201",
    "linter": "tryceratops",
    "summary": "Use `raise` without specifying exception name",
    "message_formats": [
      "Use `raise` without specifying exception name"
    ],
    "fix": "Fix is always available.",
    "explanation": "## What it does\nChecks for needless exception names in `raise` statements.\n\n## Why is this bad?\nIt's redundant to specify the exception name in a `raise` statement if the\nexception is being re-raised.\n\n## Example\n```python\ndef foo():\n    try:\n        ...\n    except ValueError as exc:\n        raise exc\n```\n\nUse instead:\n```python\ndef foo():\n    try:\n        ...\n    except ValueError:\n        raise\n```\n\n## Fix safety\nThis rule's fix is marked as unsafe, as it doesn't properly handle bound\nexceptions that are shadowed between the `except` and `raise` statements.\n",
    "preview": false
  },
...
]
```

`fix` will only be one of 3 values today. https://github.com/astral-sh/ruff/blob/f8f2e2a442aae2c1dfb4949e5ce99dc7a9d3e902/crates/ruff_diagnostics/src/violation.rs#L10-L18

## Possible options

* Update the `fix` property to add some sort of indicator on if the fix will be unsafe or not. 
* Add a new property that indicates unsafe is `true` or `false`. Similar to how `preview` is done. 
* Maybe its a new flag on the cli that `--fixable` that can be used alongside or instead of `--all`

### Keywords
unsafe fix rule property cli flag

**Comment by dhruvmanila** [comment]
Thanks for providing all the details!

> We have a legacy python project that we want to ease into using ruff and would like to start with just formatting and any linting that can be automatically fixed. Using the `ruff rule --all --output-format json` we should be able to build a command that allows for that.
> 
> Command we are using to build the list (tested only on MacOS)
> 
> ```
> ruff rule --all --output-format json | jq -r '.[] | select(.fix == "Fix is always available." and .preview == false and .code ) | .code' | sort | paste -sd, - | sed 's/[^,]*/"&"/g'
> ```
> 
> We can then just take this and put it into the `select` section of the config.

I believe this can be achieved using the following command 😅:
```
ruff check --select=ALL --fix
```

Or, am I misunderstanding what you're trying to achieve?

**Comment by dhruvmanila** [comment]
I see, thanks for providing the details. I think you already know this but not all rules provide a fix, and the `--fix` flag will apply the fix for all the selected rules that provides an autofix. You can learn more about fixes at https://docs.astral.sh/ruff/linter/#fixes.

Regardless, I don't think this is possible as fix applicability depends on the generated fix itself and is **dynamic** in a way that it could be "safe" or "unsafe" to fix the _same_ rule depending on the source code. And, the `ruff rule` command is meant to provide information about a rule irrespective of the source code. Does that make sense?

**Comment by MichaReiser** [comment]
Yes, we can't support this because the fix safety isn't static metadata. It depends on case per case, and the only way to make it static metadata would be to offer fewer fixes, which I don't think is a good outcome.
